File: ruby/triangle/triangle.rb
class Triangle
  def initialize(sides)
    @sides = sides
  end

  def equilateral?
    return false if invalid?

    unique_sides == 1
  end

  def isosceles?
    return false if invalid?

    unique_sides <= 2
  end

  def scalene?
    return false if invalid?

    unique_sides == 3
  end

  private

  attr_reader :sides

  def invalid?
    a, b, c = sides.sort
    sides.any?(&:zero?) || a + b < c
  end

  def unique_sides
    sides.uniq.length
  end
end

File: python/triangle/triangle.py
def equilateral(sides):
    if invalid(sides):
        return False

    return len(set(sides)) == 1


def isosceles(sides):
    if invalid(sides):
        return False

    return len(set(sides)) <= 2


def scalene(sides):
    pass


def invalid(sides):
    a, b, c = sorted(sides)

    return any(side == 0 for side in sides) or a + b < c
